x/application/keeper/msg_server_undelegate_from_gateway_test.go

```diff
@@ -272,3 +281,225 @@ func TestMsgServer_UndelegateFromGateway_SuccessfullyUndelegateFromUnstakedGatew
+	)
+
+	// Verify that the other gateways are still delegated to the application.
+	require.Contains(t, app.DelegateeGatewayAddresses, delegatedGatewayAddr)
+
+	// Verify that the reconstructed delegatee gateway list include the undelegated gateway.
+	gatewayAddresses := rings.GetRingAddressesAtBlock(&app, sdkCtx.BlockHeight())
+	require.Contains(t, gatewayAddresses, gatewayAddrToUndelegate)
+
+	// Increment the block height to the next session and  run the pruning
+	// undelegations logic again.
+	nextSessionStartHeight := int64(sessionEndHeight + 1)
+	sdkCtx = sdkCtx.WithBlockHeight(nextSessionStartHeight)
+	k.EndBlockerPruneAppToGatewayPendingUndelegation(sdkCtx)
+
+	// Get the updated application state.
+	app, isAppFound = k.GetApplication(sdkCtx, app.Address)
+	require.True(t, isAppFound)
+	require.NotNil(t, app)
+
+	// Verify that when queried for the next session the reconstructed delegatee
+	// gateway list does include the undelegated gateway.
+	nextSessionGatewayAddresses := rings.GetRingAddressesAtBlock(&app, nextSessionStartHeight)
+	require.NotContains(t, nextSessionGatewayAddresses, gatewayAddrToUndelegate)
+}
+
+func TestMsgServer_UndelegateFromGateway_DelegationIsPrunedAfterRetentionPeriod(t *testing.T) {
+	k, ctx := keepertest.ApplicationKeeper(t)
+	srv := keeper.NewMsgServerImpl(k)
+
+	undelegationHeight := int64(1)
+	sdkCtx, app, gatewayAddrs := delegateThenUndelegate(ctx, t, srv, k, undelegationHeight)
+
+	gatewayAddrToUndelegate := gatewayAddrs[0]
+	delegatedGatewayAddr := gatewayAddrs[1]
+
+	// Increment the block height past the undelegation retention period then run
+	// the pruning undelegations logic.
+	pruningBlockHeight := getPruningBlockHeight(undelegationHeight)
+	sdkCtx = sdkCtx.WithBlockHeight(pruningBlockHeight)
+	k.EndBlockerPruneAppToGatewayPendingUndelegation(sdkCtx)
+
+	// Get the updated application state.
+	app, isAppFound := k.GetApplication(sdkCtx, app.Address)
+	require.True(t, isAppFound)
+	require.NotNil(t, app)
+
+	// Verify that the the pending undelegation map no longer contains the
+	// sessionEndHeight key.
+	sessionEndHeight := uint64(sessionkeeper.GetSessionEndBlockHeight(undelegationHeight))
+	require.Empty(t, app.PendingUndelegations[sessionEndHeight])
+
+	// Verify that the reconstructed delegatee gateway list can no longer include
+	// the undelegated gateway since it has been pruned.
+	gatewayAddressesAfterPruning := rings.GetRingAddressesAtBlock(&app, undelegationHeight)
+	require.NotContains(t, gatewayAddressesAfterPruning, gatewayAddrToUndelegate)
+	require.Contains(t, gatewayAddressesAfterPruning, delegatedGatewayAddr)
+}
+
+func TestMsgServer_UndelegateFromGateway_RedelegationAfterUndelegationAtTheSameSessionNumber(t *testing.T) {
+	k, ctx := keepertest.ApplicationKeeper(t)
+	srv := keeper.NewMsgServerImpl(k)
+
+	undelegationHeight := int64(1)
+	sdkCtx, app, gatewayAddrs := delegateThenUndelegate(ctx, t, srv, k, undelegationHeight)
+
+	gatewayAddrToRedelegate := gatewayAddrs[0]
```

Comment:
There's a lot of implicit knowledge here that index 0 is the one that was undelegated above.

Can you change the helper function to return `ctx, app, delegatedToAddr, pendingUndelegateToAddr` so it's VERY clear and explicit.